### ISSUE TITLE
vocabulary builder 

### ISSUE THREAD

**Issue by keighrim** [clamsproject/mmif#7]
The goal of developing vocab builder is to automate publication of vocabulary as a website. The raw definition of the vocabulary itself  (annotation types, object features, ...) must be written in commonly used data format (such as json, yaml, etc) so that anyone can send pull requests for updating the vocab for addition of new types of annotations (and we can use comments in the PR for discussion). The bulider should take a single definition file or a single directory of those files to build HTML files and copy them to `docs` directory so that github will automatically publish them via http://clamsproject.github.io/mmif. Current prototype (available in [`vocabulary`](https://github.com/clamsproject/mmif/tree/vocabulary) branch) is using yaml as input.

**Comment by marcverhagen** [comment]
Current version takes a single file and writes a directory with a version number to `docs/vocabulary`, but it does not create RDF, OWL or TTL like with the LAPPS vocabulary.

**Comment by keighrim** [comment]
This issue now is a subtask of #19.

**Comment by keighrim** [comment]
As outlined in #19 , the `vocab` builder should be able to perform these two actions at least.
* `make publish` - when there's an update in `vocab`, update `context` and `sdk` accordingly and version everything, update links in `spec`, and publish
* `make dev` - do the same as make publish but no publishing (no push to github in essence)

Concretely, 
* `make dev`
  1. ask user for a new version number (or ask if this upgrade is *major*, *minor*, or *patch*, see #14)
  1. bump up version number in `VERSION` file and by git-tagging
  1. build HTML, OWL, TTL (or RDF?) from new `vocab` (written in YAML)
    * can be supplemented with static css, js files if needed
  1. copy generated files (and static web files, if any) to `docs` for github.io publication
* `make publish` 
  1. do everything in `make dev`
  1. add files in `docs` and create a commit
  1. push to github (commits and a new tag)

**Comment by marcverhagen** [comment]
Will committing and pushing publish? I think at the moment nothing is published until you merge into master. I think that is nice because it adds a step of reflection where we create a pull request and sit on it for a moment.

**Comment by keighrim** [comment]
@marcverhagen yeah, that makes much more sense. So it should push at the end in a new branch and create a PR, I guess.

**Comment by marcverhagen** [comment]
@keighrim I would keep that last part fully manual. So we commit and push on a develop or version branch and automate all that, but leave the last step of creating a pull request a manual one, it is not something we would do more than once a month at most.

**Comment by keighrim** [comment]
One problem is that in this way, the git tag might live in the non-main branch. So instead of having git-tagging as a part of builder, we might want to use post-merge hook to tag the main branch base don `VERSION` file.

**Comment by marcverhagen** [comment]
Ah, we don't want that, tags outside of master. I was thinking that once we are ready to create a new version we branch off from develop or whatnot and create a branch version-Major-Minor-Patch, merge that into master when ready and then create the tag. I don't think there is any need to create branches automatically.

On the `VERSION` file. It has the version that we want to publish. In your approach it should never be edited manually since the system depends on knowing what version we are updating from. We should be disciplined on that. And this does also mean that we need to be careful when we commit twice. Say we are at version 1.2.0 and are working on a minor update. If with a `make dev` the `VERSION` file is changed then we want to change it only once, so the next time around the user should not say "minor update" again because then we would go from 1.2.0 to 1.3.0 and so on. So updating `VERSION` should be done by your first option, asking for the number.

**Comment by keighrim** [comment]
#39 adds ability to generate PDF based file formats.

**Comment by keighrim** [comment]
As 77bc9be dropped the PDF/TTL generation (see #89 for related discussion), the wanted features of this *builder* are pretty much all available in `publish.py`. I think we can close this issue.

**Comment by keighrim** [comment]
With the separation of `mmif-python` from MMIF spec, I think current `build.py` (formerly `publish.py`) is doing everythin we want from a *vocabulary builder*. And thus, I believe this issue can be closed.

**Comment by marcverhagen** [comment]
Yes, I agree. I will make some small changes when building the next version, but we do not need to keep this issue open for that.
